### ISSUE TITLE
Loss Functions.... Dice score in Inference

### ISSUE THREAD

**Issue by imruljubair** [MIC-DKFZ/nnUNet#1792]
Hi,
Thank you for developing this amazing technique and sharing with us.

Note: I am new to nnUNet and also in medical images. I apologize if my questions do not make sense.... :(

Questions:
(1) What is the meaning of _pseudo dice_ and _pseudo dice (mov. avg.)_ in the progress.png
(2) Can I change training loss from _Dice+CE_ to _Dice or General Dice_ only for experimental purposes? Where should I change it?
(3) Can I use Dice score only during inference to measure the performance?

The reason I am trying incorporate Dice score is: most of the paper I trying to compare with used only Dice score in their experiments. To be on the same page, I also need to apply nnUNet with Dice score, not Dice+CE.

**Comment by mrokuss** [comment]
Hi @imruljubair,

Thank you for your questions, I'll address them separately:

1.  **Pseudo Dice and Pseudo Dice (mov. avg.) in progress.png**:  The "pseudo" in pseudo dice refers to the fact that the dice scores are computed on the patches (parts of the images) the network sees during training. The pseudo dice score is a surrogate for the actual dice score and is used during training to monitor the network's progress. The "mov. avg." stands for moving average, which is a smoothed version of the dice score over time, helping to visualize trends more easily.
2. **Changing training loss from Dice+CE to Dice:** Yes, you can modify the training loss for your experimental purposes. Just change the nnUNetTrainer to [nnUNetTrainerDiceLoss](https://github.com/MIC-DKFZ/nnUNet/blob/2bc504de420fb64f130bfe22ce9867cbad2b9d4a/nnunetv2/training/nnUNetTrainer/variants/loss/nnUNetTrainerDiceLoss.py#L11). Make sure to refer to the nnUNet documentation for the exact syntax and options: `nnUNetv2_train DATASET_NAME_OR_ID UNET_CONFIGURATION FOLD -tr nnUNetTrainerDiceLoss`
3.  Using Dice score only during inference: Yes, you can certainly use the Dice score alone during inference to evaluate the performance of your model. During inference, the trained model generates predictions, and you can compute the Dice score on these predictions compared to the ground truth masks. For the 5-fold cross validation this is already done for you and you can find the results in the `summary.json`.


Remember to consult the nnUNet documentation for information on configuring and customizing nnUNet for your experiments. Good luck with your work! Best,

Max

**Comment by imruljubair** [comment]
Hi @mrokuss ,

Thank you so much for such a clear explanation. It makes a lot senses.

I just have a follow-up query on "**Using Dice score only during inference**":
Could you suggest anything from your code that I can use to compare test cases with ground truths based on Dice scores? Is it the `MemoryEfficientSoftDiceLoss()`? I would like to reuse your Dice implementation to compare test and ground truth, not from some random Dice code from internet... :)

**Comment by mrokuss** [comment]
Hey @imruljubair 

The `MemoryEfficientSoftDiceLoss()`  is a Loss function used during training to compute the Dice on the labels and softmax output. What you want is an implementation of the dice metric. The calculation of all metrics found in the `summary.json` can be found [here]( https://github.com/MIC-DKFZ/nnUNet/blob/b00b41f240540bac12911519c05fcc8012e594d7/nnunetv2/evaluation/evaluate_predictions.py#L89).

Best,

Max
